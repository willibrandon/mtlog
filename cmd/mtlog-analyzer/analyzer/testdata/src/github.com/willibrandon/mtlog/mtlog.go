// Stub mtlog package for tests
package mtlog

// Logger is a stub for the mtlog logger
type Logger struct{}

// New creates a new logger
func New() *Logger { return &Logger{} }

// Debug logs at debug level
func (l *Logger) Debug(template string, args ...interface{}) {}

// Information logs at information level  
func (l *Logger) Information(template string, args ...interface{}) {}

// Warning logs at warning level
func (l *Logger) Warning(template string, args ...interface{}) {}

// Error logs at error level
func (l *Logger) Error(template string, args ...interface{}) {}

// Fatal logs at fatal level
func (l *Logger) Fatal(template string, args ...interface{}) {}

// Verbose logs at verbose level
func (l *Logger) Verbose(template string, args ...interface{}) {}

// Short method names
func (l *Logger) V(template string, args ...interface{}) {}
func (l *Logger) D(template string, args ...interface{}) {}
func (l *Logger) I(template string, args ...interface{}) {}
func (l *Logger) W(template string, args ...interface{}) {}
func (l *Logger) E(template string, args ...interface{}) {}
func (l *Logger) F(template string, args ...interface{}) {}

// ForContext returns a logger for the given context
func (l *Logger) ForContext(key string, value interface{}) *Logger { return l }