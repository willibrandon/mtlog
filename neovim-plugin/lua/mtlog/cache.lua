-- Cache module for mtlog.nvim

local M = {}

local config = require('mtlog.config')
local utils = require('mtlog.utils')

-- Cache storage
local cache = {}

-- Cache statistics
local stats = {
  hits = 0,
  misses = 0,
  size = 0,
}

-- LRU tracking
local lru_order = {} -- List of keys in LRU order (most recent at end)

-- Analyzer version for cache invalidation
local analyzer_version = nil

-- Cache entry structure
---@class CacheEntry
---@field data any Cached data
---@field mtime number File modification time
---@field version string Analyzer version
---@field timestamp number Cache timestamp

-- Get cache key for file
---@param filepath string File path
---@return string Cache key
local function get_cache_key(filepath)
  return vim.fn.fnamemodify(filepath, ':p')
end

-- Update LRU order for a key
---@param key string Cache key
local function update_lru(key)
  -- Remove from current position if exists
  for i, k in ipairs(lru_order) do
    if k == key then
      table.remove(lru_order, i)
      break
    end
  end
  -- Add to end (most recent)
  table.insert(lru_order, key)
end

-- Evict LRU entries if max size exceeded
local function enforce_max_size()
  local max_size = config.get('cache.max_size')
  if not max_size or max_size <= 0 then
    return
  end
  
  while #lru_order > max_size do
    local oldest_key = table.remove(lru_order, 1)
    cache[oldest_key] = nil
    stats.size = stats.size - 1
  end
end

-- Get current analyzer version
---@return string Version string
local function get_analyzer_version()
  if analyzer_version then
    return analyzer_version
  end
  
  -- Get version from analyzer module
  local analyzer = require('mtlog.analyzer')
  analyzer_version = analyzer.get_version() or 'unknown'
  
  return analyzer_version
end

-- Check if cache entry is valid
---@param entry CacheEntry Cache entry
---@param filepath string File path
---@return boolean Is valid
local function is_cache_valid(entry, filepath)
  -- Check if cache is enabled
  if not config.get('cache.enabled') then
    return false
  end
  
  -- Check analyzer version
  if entry.version ~= get_analyzer_version() then
    return false
  end
  
  -- Check file modification time
  local current_mtime = utils.get_mtime(filepath)
  if not current_mtime or current_mtime ~= entry.mtime then
    return false
  end
  
  -- Check TTL
  local ttl = config.get('cache.ttl_seconds')
  if ttl and ttl > 0 then
    local age = os.time() - entry.timestamp
    if age > ttl then
      return false
    end
  end
  
  return true
end

-- Get cached data
---@param filepath string File path
---@return any? Cached data or nil
function M.get(filepath)
  if not config.get('cache.enabled') then
    return nil
  end
  
  local key = get_cache_key(filepath)
  local entry = cache[key]
  
  if not entry then
    stats.misses = stats.misses + 1
    return nil
  end
  
  -- Validate cache entry
  if not is_cache_valid(entry, filepath) then
    cache[key] = nil
    stats.size = stats.size - 1
    stats.misses = stats.misses + 1
    -- Remove from LRU
    for i, k in ipairs(lru_order) do
      if k == key then
        table.remove(lru_order, i)
        break
      end
    end
    return nil
  end
  
  stats.hits = stats.hits + 1
  update_lru(key)
  return entry.data
end

-- Set cached data
---@param filepath string File path
---@param data any Data to cache
function M.set(filepath, data)
  if not config.get('cache.enabled') then
    return
  end
  
  local key = get_cache_key(filepath)
  local mtime = utils.get_mtime(filepath)
  
  if not mtime then
    return
  end
  
  -- Check if updating existing entry
  local is_new = cache[key] == nil
  
  cache[key] = {
    data = data,
    mtime = mtime,
    version = get_analyzer_version(),
    timestamp = os.time(),
  }
  
  if is_new then
    stats.size = stats.size + 1
  end
  
  update_lru(key)
  enforce_max_size()
end

-- Invalidate cache for file
---@param filepath string File path
function M.invalidate(filepath)
  local key = get_cache_key(filepath)
  cache[key] = nil
end

-- Clear entire cache
function M.clear()
  cache = {}
  lru_order = {}
  stats.size = 0
  -- Don't reset hits/misses unless explicitly asked
end

-- Clear expired entries
function M.cleanup()
  local ttl = config.get('cache.ttl_seconds')
  if not ttl or ttl <= 0 then
    return
  end
  
  local now = os.time()
  local expired = {}
  
  for key, entry in pairs(cache) do
    if (now - entry.timestamp) > ttl then
      table.insert(expired, key)
    end
  end
  
  for _, key in ipairs(expired) do
    cache[key] = nil
    stats.size = stats.size - 1
    -- Remove from LRU order
    for i, k in ipairs(lru_order) do
      if k == key then
        table.remove(lru_order, i)
        break
      end
    end
  end
end

-- Get cache statistics
---@return table Statistics
function M.stats()
  return {
    size = stats.size,
    hits = stats.hits,
    misses = stats.misses,
    hit_rate = stats.hits > 0 and (stats.hits / (stats.hits + stats.misses)) or 0,
    entries = stats.size,  -- Alias for compatibility
  }
end

-- Reset cache statistics
function M.reset_stats()
  stats.hits = 0
  stats.misses = 0
  -- Don't reset size as it's tracked automatically
end

-- Invalidate cache on analyzer version change
function M.invalidate_on_version_change()
  local new_version = get_analyzer_version()
  
  if analyzer_version and analyzer_version ~= new_version then
    -- Version changed, clear cache
    M.clear()
  end
  
  analyzer_version = new_version
end

-- Save cache to disk (for persistence across sessions)
---@param filepath string Path to cache file
---@return boolean Success
function M.save_to_disk(filepath)
  -- Create cache directory
  local cache_dir = vim.fn.fnamemodify(filepath, ':h')
  if not utils.ensure_directory(cache_dir) then
    return false
  end
  
  -- Serialize cache
  local data = vim.json.encode(cache)
  if not data then
    return false
  end
  
  -- Write to file
  return utils.write_file(filepath, data)
end

-- Load cache from disk
---@param filepath string Path to cache file
---@return boolean Success
function M.load_from_disk(filepath)
  -- Read file
  local data = utils.read_file(filepath)
  if not data then
    return false
  end
  
  -- Deserialize cache
  local ok, loaded = pcall(vim.json.decode, data)
  if not ok or type(loaded) ~= 'table' then
    return false
  end
  
  -- Validate and load entries
  for key, entry in pairs(loaded) do
    if type(entry) == 'table' and entry.data and entry.mtime and entry.version and entry.timestamp then
      cache[key] = entry
    end
  end
  
  -- Clean up expired entries
  M.cleanup()
  
  return true
end

-- Get cache file path
---@return string Cache file path
function M.get_cache_file()
  local cache_dir = vim.fn.stdpath('cache')
  return cache_dir .. '/mtlog-analyzer-cache.json'
end

-- Setup automatic cache management
function M.setup()
  -- Load cache from disk on startup
  local cache_file = M.get_cache_file()
  if vim.fn.filereadable(cache_file) == 1 then
    M.load_from_disk(cache_file)
  end
  
  -- Set up periodic cleanup
  local cleanup_timer = vim.loop.new_timer()
  cleanup_timer:start(60000, 60000, vim.schedule_wrap(function()
    M.cleanup()
  end))
  
  -- Save cache on exit
  vim.api.nvim_create_autocmd('VimLeavePre', {
    callback = function()
      M.save_to_disk(cache_file)
    end,
  })
  
  -- Invalidate cache on analyzer update
  vim.api.nvim_create_autocmd('User', {
    pattern = 'MtlogAnalyzerUpdated',
    callback = function()
      M.invalidate_on_version_change()
    end,
  })
end

return M