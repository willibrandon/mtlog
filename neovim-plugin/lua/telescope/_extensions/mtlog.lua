local has_telescope, telescope = pcall(require, 'telescope')
if not has_telescope then
  error('This extension requires telescope.nvim')
end

local actions = require('telescope.actions')
local action_state = require('telescope.actions.state')
local conf = require('telescope.config').values
local finders = require('telescope.finders')
local pickers = require('telescope.pickers')
local previewers = require('telescope.previewers')
local entry_display = require('telescope.pickers.entry_display')
local utils = require('telescope.utils')

local M = {}

-- Severity icons and highlights
local severity_icons = {
  [vim.diagnostic.severity.ERROR] = { icon = ' ', hl = 'DiagnosticError' },
  [vim.diagnostic.severity.WARN] = { icon = ' ', hl = 'DiagnosticWarn' },
  [vim.diagnostic.severity.INFO] = { icon = ' ', hl = 'DiagnosticInfo' },
  [vim.diagnostic.severity.HINT] = { icon = ' ', hl = 'DiagnosticHint' },
}

-- Get all mtlog diagnostics from all buffers
local function get_all_diagnostics(opts)
  opts = opts or {}
  local diagnostics = {}
  local ns = vim.api.nvim_create_namespace('mtlog-analyzer')
  
  -- Get all loaded buffers
  local buffers = vim.api.nvim_list_bufs()
  
  for _, bufnr in ipairs(buffers) do
    if vim.api.nvim_buf_is_loaded(bufnr) then
      local bufname = vim.api.nvim_buf_get_name(bufnr)
      
      -- Only process Go files
      if bufname:match('%.go$') then
        local buf_diagnostics = vim.diagnostic.get(bufnr, {
          namespace = ns,
          severity = opts.severity
        })
        
        for _, diagnostic in ipairs(buf_diagnostics) do
          table.insert(diagnostics, {
            bufnr = bufnr,
            bufname = bufname,
            lnum = diagnostic.lnum + 1,  -- Convert to 1-based
            col = diagnostic.col + 1,     -- Convert to 1-based
            text = diagnostic.message,
            severity = diagnostic.severity,
            code = diagnostic.code,
            source = diagnostic.source or 'mtlog',
          })
        end
      end
    end
  end
  
  -- Sort by filename and line number
  table.sort(diagnostics, function(a, b)
    if a.bufname ~= b.bufname then
      return a.bufname < b.bufname
    end
    return a.lnum < b.lnum
  end)
  
  return diagnostics
end

-- Create the telescope picker
M.diagnostics = function(opts)
  opts = opts or {}
  
  local diagnostics = get_all_diagnostics(opts)
  
  if #diagnostics == 0 then
    vim.notify('No mtlog diagnostics found', vim.log.levels.INFO)
    return
  end
  
  -- Create displayer
  local displayer = entry_display.create({
    separator = ' ',
    items = {
      { width = 2 },  -- Icon
      { width = 20 }, -- File
      { width = 5 },  -- Line:Col
      { width = 8 },  -- Code
      { remaining = true }, -- Message
    },
  })
  
  -- Create finder
  local finder = finders.new_table({
    results = diagnostics,
    entry_maker = function(diagnostic)
      local filename = vim.fn.fnamemodify(diagnostic.bufname, ':t')
      local severity_info = severity_icons[diagnostic.severity] or { icon = '?', hl = 'Normal' }
      
      return {
        value = diagnostic,
        display = function(entry)
          return displayer({
            { severity_info.icon, severity_info.hl },
            { filename, 'TelescopeResultsField' },
            { string.format('%d:%d', diagnostic.lnum, diagnostic.col), 'TelescopeResultsNumber' },
            { diagnostic.code or '', 'TelescopeResultsComment' },
            { diagnostic.text, 'TelescopeResultsString' },
          })
        end,
        ordinal = string.format('%s %d %s', diagnostic.bufname, diagnostic.lnum, diagnostic.text),
        filename = diagnostic.bufname,
        lnum = diagnostic.lnum,
        col = diagnostic.col,
      }
    end,
  })
  
  -- Create picker
  pickers.new(opts, {
    prompt_title = 'mtlog Diagnostics',
    finder = finder,
    sorter = conf.generic_sorter(opts),
    previewer = conf.grep_previewer(opts),
    attach_mappings = function(prompt_bufnr, map)
      actions.select_default:replace(function()
        actions.close(prompt_bufnr)
        local selection = action_state.get_selected_entry()
        if selection then
          local diagnostic = selection.value
          -- Open the file and jump to location
          vim.cmd('edit ' .. diagnostic.bufname)
          vim.api.nvim_win_set_cursor(0, { diagnostic.lnum, diagnostic.col - 1 })
        end
      end)
      
      -- Add quick fix action
      map('i', '<C-q>', function()
        local selection = action_state.get_selected_entry()
        if selection then
          local diagnostic = selection.value
          -- Apply quick fix if available
          vim.cmd('edit ' .. diagnostic.bufname)
          vim.api.nvim_win_set_cursor(0, { diagnostic.lnum, diagnostic.col - 1 })
          vim.cmd('MtlogQuickFix')
        end
        actions.close(prompt_bufnr)
      end)
      
      return true
    end,
  }):find()
end

-- Picker for managing suppressed diagnostics
M.suppressions = function(opts)
  opts = opts or {}
  
  local config = require('mtlog.config')
  local mtlog_utils = require('mtlog.utils')
  local suppressed = config.get('suppressed_diagnostics') or {}
  
  if #suppressed == 0 then
    vim.notify('No diagnostics are currently suppressed', vim.log.levels.INFO)
    return
  end
  
  -- Create displayer
  local displayer = entry_display.create({
    separator = ' ',
    items = {
      { width = 10 }, -- Diagnostic ID
      { remaining = true }, -- Description
    },
  })
  
  -- Build list of suppressed diagnostics with descriptions
  local items = {}
  for _, id in ipairs(suppressed) do
    table.insert(items, {
      id = id,
      description = mtlog_utils.get_diagnostic_description(id),
    })
  end
  
  -- Create finder
  local finder = finders.new_table({
    results = items,
    entry_maker = function(item)
      return {
        value = item,
        display = function(entry)
          return displayer({
            { item.id, 'TelescopeResultsConstant' },
            { item.description, 'TelescopeResultsString' },
          })
        end,
        ordinal = item.id .. ' ' .. item.description,
      }
    end,
  })
  
  -- Create picker
  pickers.new(opts, {
    prompt_title = 'Suppressed Diagnostics',
    finder = finder,
    sorter = conf.generic_sorter(opts),
    attach_mappings = function(prompt_bufnr, map)
      -- Default action: unsuppress the selected diagnostic
      actions.select_default:replace(function()
        local selection = action_state.get_selected_entry()
        if selection then
          actions.close(prompt_bufnr)
          require('mtlog').unsuppress_diagnostic(selection.value.id)
        end
      end)
      
      -- Multi-select to unsuppress multiple
      map('i', '<Tab>', actions.toggle_selection + actions.move_selection_worse)
      map('i', '<S-Tab>', actions.toggle_selection + actions.move_selection_better)
      
      -- Unsuppress all selected
      map('i', '<CR>', function()
        local picker = action_state.get_current_picker(prompt_bufnr)
        local multi_selections = picker:get_multi_selection()
        
        actions.close(prompt_bufnr)
        
        if #multi_selections > 0 then
          for _, selection in ipairs(multi_selections) do
            require('mtlog').unsuppress_diagnostic(selection.value.id)
          end
        else
          local selection = action_state.get_selected_entry()
          if selection then
            require('mtlog').unsuppress_diagnostic(selection.value.id)
          end
        end
      end)
      
      -- Clear all suppressions
      map('i', '<C-a>', function()
        actions.close(prompt_bufnr)
        require('mtlog').unsuppress_all()
      end)
      
      return true
    end,
  }):find()
end

-- Picker for viewing workspace analysis results
M.workspace = function(opts)
  opts = opts or {}
  
  -- Run workspace analysis first
  local analyzer = require('mtlog.analyzer')
  local results = analyzer.analyze_workspace_sync()
  
  if not results or vim.tbl_isempty(results) then
    vim.notify('No issues found in workspace', vim.log.levels.INFO)
    return
  end
  
  -- Flatten results into diagnostic list
  local diagnostics = {}
  for file, file_diagnostics in pairs(results) do
    for _, diag in ipairs(file_diagnostics) do
      table.insert(diagnostics, vim.tbl_extend('force', diag, {
        bufname = file,
      }))
    end
  end
  
  -- Sort by severity, then file, then line
  table.sort(diagnostics, function(a, b)
    if a.severity ~= b.severity then
      return a.severity < b.severity  -- ERROR < WARN < INFO < HINT
    end
    if a.bufname ~= b.bufname then
      return a.bufname < b.bufname
    end
    return a.lnum < b.lnum
  end)
  
  -- Use the same picker as diagnostics
  opts.prompt_title = 'mtlog Workspace Analysis'
  M.diagnostics(vim.tbl_extend('force', opts, { _diagnostics = diagnostics }))
end

-- Register the extension
return telescope.register_extension({
  setup = function(ext_config, config)
    -- Extension setup
  end,
  exports = {
    mtlog = M.diagnostics,
    diagnostics = M.diagnostics,
    workspace = M.workspace,
    suppressions = M.suppressions,
  },
})