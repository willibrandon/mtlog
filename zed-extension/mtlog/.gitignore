# Rust build artifacts
/target/
/Cargo.lock

# IDE and editor files
.DS_Store
*.swp
*.swo
*~
.idea/
.vscode/

# Debug files
*.pdb
*.dSYM/

# Backup files
*.bak
*.orig

# WASM specific
*.wasm.map

# Temporary files
tmp/
temp/
*.tmp