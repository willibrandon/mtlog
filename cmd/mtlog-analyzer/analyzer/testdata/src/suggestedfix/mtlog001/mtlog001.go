package mtlog001

import "github.com/willibrandon/mtlog"

func testTemplateMismatchFixes() {
	log := mtlog.New()

	// Too few arguments - should suggest adding placeholders
	log.Information("User {UserId} performed {Action} on {Resource}") // want `\[MTLOG001\] template has 3 properties but 0 arguments provided`
	
	// Too few arguments with existing comment - TODO should go on next line
	log.Information("Failed {Operation}") // existing comment // want `\[MTLOG001\] template has 1 properties but 0 arguments provided`

	// Too many arguments - should suggest removing extras
	log.Information("Hello {Name}", "Alice", "Bob", "Charlie") // want `\[MTLOG001\] template has 1 properties but 3 arguments provided`

	// Correct number of arguments - no fix needed
	log.Information("User {UserId} logged in", 123)
}