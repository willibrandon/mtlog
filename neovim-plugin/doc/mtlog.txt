*mtlog.txt*    mtlog-analyzer integration for Neovim

================================================================================
CONTENTS                                                         *mtlog-contents*

    1. Introduction ......................................... |mtlog-intro|
    2. Requirements ......................................... |mtlog-requirements|
    3. Installation ......................................... |mtlog-installation|
    4. Configuration ........................................ |mtlog-config|
    5. Commands ............................................. |mtlog-commands|
    6. Features ............................................. |mtlog-features|
        6.1 Diagnostics ..................................... |mtlog-diagnostics|
        6.2 Quick Fixes ..................................... |mtlog-quickfixes|
        6.3 Suppressions .................................... |mtlog-suppressions|
        6.4 LSP Integration ................................. |mtlog-lsp|
        6.5 Context Rules ................................... |mtlog-context|
        6.6 Queue Management ................................ |mtlog-queue|
    7. Telescope Integration ................................ |mtlog-telescope|
    8. Statusline Integration ............................... |mtlog-statusline|
    9. Troubleshooting ...................................... |mtlog-troubleshooting|
    10. About ............................................... |mtlog-about|

================================================================================
INTRODUCTION                                                        *mtlog-intro*

mtlog.nvim provides seamless integration with mtlog-analyzer, a static analysis
tool for the mtlog structured logging library. It catches common logging
mistakes at development time, improving code quality and preventing runtime
errors.

Key benefits:
• Real-time validation of message templates
• Automatic quick fixes for common issues
• LSP integration for native code actions
• Smart context-aware analysis control
• Efficient queue-based processing

================================================================================
REQUIREMENTS                                                 *mtlog-requirements*

• Neovim >= 0.8.0
• Go project using the mtlog library
• mtlog-analyzer installed (install with: go install github.com/willibrandon/mtlog/cmd/mtlog-analyzer@latest)
• Optional: Telescope.nvim for enhanced UI
• Optional: gopls for full LSP integration

================================================================================
INSTALLATION                                                 *mtlog-installation*

Using lazy.nvim: >lua
    {
      'willibrandon/mtlog',
      ft = 'go',
      dir = vim.fn.expand('~/path/to/mtlog/neovim-plugin'),
      config = function()
        require('mtlog').setup({
          -- your configuration
        })
      end,
      dependencies = {
        'nvim-telescope/telescope.nvim', -- optional
      },
    }
<

Using packer.nvim: >lua
    use {
      '~/path/to/mtlog/neovim-plugin',
      ft = 'go',
      config = function()
        require('mtlog').setup()
      end,
    }
<

================================================================================
CONFIGURATION                                                     *mtlog-config*

Setup function accepts these options: >lua
    require('mtlog').setup({
      -- Analyzer settings
      analyzer_path = 'mtlog-analyzer',       -- Path to analyzer binary
      analyzer_flags = {},                    -- Additional flags
      
      -- Auto-enable for Go projects
      auto_enable = true,
      auto_analyze = true,
      
      -- Debouncing
      debounce_ms = 250,
      
      -- Diagnostics
      diagnostics_enabled = true,             -- Global kill switch
      show_diagnostics = true,
      virtual_text = true,
      signs = true,
      underline = true,
      update_in_insert = false,
      severity_sort = true,
      
      -- Suppressed diagnostics (e.g., {'MTLOG001', 'MTLOG005'})
      suppressed_diagnostics = {},
      
      -- LSP Integration
      lsp_integration = {
        enabled = true,                      -- Enable fake LSP client
        show_suppress_action = true,         -- Show suppress in code actions
      },
      
      -- Context rules for auto-enable/disable
      context_rules = {},                    -- Custom rules
      use_builtin_rules = true,             -- Use default rules
      
      -- Queue settings
      queue = {
        max_concurrent = nil,                -- Auto-detect from CPU
        timeout_ms = 10000,
      },
    })
<

Context Rules Example: >lua
    context_rules = {
      {
        type = 'path',
        pattern = '_test%.go$',              -- Test files
        action = 'disable',
        description = 'Disable for test files',
      },
      {
        type = 'buffer',
        line_count = { max = 1000 },         -- Large files
        action = 'disable',
        description = 'Disable for large files',
      },
      {
        type = 'project',
        has_file = 'go.mod',                 -- Go modules
        action = 'enable',
        description = 'Enable for Go modules',
      },
    }
<

================================================================================
COMMANDS                                                       *mtlog-commands*

                                                               *:MtlogAnalyze*
:MtlogAnalyze [file]
    Run analyzer on current buffer or specified file.

                                                       *:MtlogAnalyzeWorkspace*
:MtlogAnalyzeWorkspace
    Analyze all Go files in the workspace.

                                                                 *:MtlogClear*
:MtlogClear[!]
    Clear diagnostics for current buffer. Use ! to clear all buffers.

                                                                *:MtlogEnable*
:MtlogEnable
    Enable the analyzer for the current session.

                                                               *:MtlogDisable*
:MtlogDisable
    Disable the analyzer for the current session.

                                                                *:MtlogToggle*
:MtlogToggle
    Toggle analyzer on/off.

                                                     *:MtlogToggleDiagnostics*
:MtlogToggleDiagnostics
    Toggle the global diagnostics kill switch.

                                                              *:MtlogSuppress*
:MtlogSuppress [diagnostic_id]
    Suppress a diagnostic ID (e.g., MTLOG001).

                                                            *:MtlogUnsuppress*
:MtlogUnsuppress [diagnostic_id]
    Remove suppression for a diagnostic ID.

                                                         *:MtlogUnsuppressAll*
:MtlogUnsuppressAll
    Clear all diagnostic suppressions.

                                                       *:MtlogShowSuppressions*
:MtlogShowSuppressions
    Display currently suppressed diagnostics.

                                                     *:MtlogManageSuppressions*
:MtlogManageSuppressions
    Open Telescope picker to manage suppressions.

                                                             *:MtlogWorkspace*
:MtlogWorkspace {save|load|path}
    Manage workspace configuration (.mtlog.json).

                                                                *:MtlogStatus*
:MtlogStatus
    Show plugin status and diagnostic counts.

                                                                 *:MtlogCache*
:MtlogCache {clear|stats}
    Manage the analysis cache.

                                                               *:MtlogContext*
:MtlogContext {show|test|add-builtin|clear}
    Manage context rules for automatic control.

                                                                 *:MtlogQueue*
:MtlogQueue {show|stats|clear|pause|resume}
    Manage the analysis queue.

                                                            *:MtlogCodeAction*
:MtlogCodeAction
    Show mtlog code actions at cursor.

                                                              *:MtlogQuickFix*
:MtlogQuickFix
    Apply quick fix at cursor position.

================================================================================
FEATURES                                                       *mtlog-features*

--------------------------------------------------------------------------------
DIAGNOSTICS                                                  *mtlog-diagnostics*

The plugin detects these diagnostic codes:

• MTLOG001: Template/argument mismatch
• MTLOG002: Invalid format specifier
• MTLOG003: Missing error in Error/Fatal calls
• MTLOG004: Property name should be PascalCase
• MTLOG005: Duplicate property in template
• MTLOG006: Type should implement LogValue
• MTLOG007: String should be constant
• MTLOG008: Dynamic template warning
• MTLOG009: With() odd argument count (key-value pairs)
• MTLOG010: With() non-string key
• MTLOG011: With() cross-call duplicate property
• MTLOG012: With() reserved property name
• MTLOG013: With() empty key string

Diagnostics appear as:
• Virtual text (configurable)
• Signs in the sign column
• Underlines in the buffer
• Entries in the quickfix list

--------------------------------------------------------------------------------
QUICK FIXES                                                  *mtlog-quickfixes*

Many diagnostics come with automatic fixes:

• MTLOG001: Add missing arguments or remove extra ones
• MTLOG002: Correct format specifiers
• MTLOG003: Add error parameter
• MTLOG004: Convert to PascalCase
• MTLOG005: Remove duplicate properties
• MTLOG007: Extract string to constant
• MTLOG009: Add missing value or remove dangling key
• MTLOG010: Convert non-string key to string
• MTLOG012: Suggest alternative property name
• MTLOG013: Remove empty key-value pair

Apply fixes:
• `:MtlogQuickFix` at cursor position
• `:lua vim.lsp.buf.code_action()` for LSP integration
• `<leader>mf` with recommended mappings

--------------------------------------------------------------------------------
SUPPRESSIONS                                                *mtlog-suppressions*

Suppress diagnostics you don't want to see:

Session suppression: >vim
    :MtlogSuppress MTLOG004
<

Workspace suppression (saved to .mtlog.json): >vim
    :MtlogSuppress MTLOG004
    " Choose 'Yes' when prompted to save to workspace
<

Environment variable suppression: >bash
    export MTLOG_SUPPRESS="MTLOG004,MTLOG008"
<

Inline suppression in code: >go
    // mtlog:disable-next-line MTLOG004
    log.Info("message with {snake_case}", value)
<

--------------------------------------------------------------------------------
LSP INTEGRATION                                                     *mtlog-lsp*

The plugin creates a fake LSP client to integrate with Neovim's built-in LSP
features. This allows mtlog code actions to appear alongside gopls actions.

Enable in config: >lua
    lsp_integration = {
      enabled = true,
      show_suppress_action = true,
    }
<

Use standard LSP commands: >vim
    :lua vim.lsp.buf.code_action()
<

The mtlog actions will appear in the code action menu with gopls and other
LSP providers.

--------------------------------------------------------------------------------
CONTEXT RULES                                                   *mtlog-context*

Context rules automatically enable/disable analysis based on file properties.
This feature is unique to the Neovim plugin.

Rule types:
• `path` - Match file paths with patterns
• `buffer` - Match buffer properties (line count, size)
• `project` - Match project properties (has_file, has_dir)
• `custom` - Custom Lua function

Example rules: >lua
    context_rules = {
      -- Disable for test files
      {
        type = 'path',
        pattern = '_test%.go$',
        action = 'disable',
      },
      -- Disable for generated files
      {
        type = 'path',
        pattern = '%.pb%.go$',
        action = 'disable',
      },
      -- Disable for large files
      {
        type = 'buffer',
        line_count = { max = 1000 },
        action = 'disable',
      },
      -- Custom function
      {
        type = 'custom',
        check = function(bufnr)
          return vim.bo[bufnr].readonly
        end,
        action = 'disable',
      },
    }
<

Built-in rules available via `:MtlogContext add-builtin`.

--------------------------------------------------------------------------------
QUEUE MANAGEMENT                                                   *mtlog-queue*

The plugin uses a priority queue to manage analysis tasks efficiently:

Priority levels:
• HIGH (1) - Current buffer changes
• NORMAL (2) - Visible buffers
• LOW (3) - Background/workspace analysis

Features:
• CPU-aware concurrency limits
• Automatic deduplication
• Graceful cancellation
• Progress tracking

Monitor with `:MtlogQueue show` or `:MtlogQueue stats`.

================================================================================
TELESCOPE INTEGRATION                                         *mtlog-telescope*

If Telescope is installed, additional pickers are available:

                                                    *telescope.mtlog.diagnostics*
:Telescope mtlog diagnostics
    Browse and jump to diagnostics across all buffers.

                                                   *telescope.mtlog.suppressions*
:Telescope mtlog suppressions
    Manage suppressed diagnostics with preview.

                                                      *telescope.mtlog.quickfixes*
:Telescope mtlog quickfixes
    Browse available quick fixes for current buffer.

Setup in Telescope config: >lua
    require('telescope').load_extension('mtlog')
<

================================================================================
STATUSLINE INTEGRATION                                       *mtlog-statusline*

Add diagnostic counts to your statusline:

For lualine: >lua
    sections = {
      lualine_c = {
        {
          require('mtlog.statusline').component,
          cond = require('mtlog.statusline').has_diagnostics,
        },
      },
    }
<

For custom statuslines: >lua
    local counts = require('mtlog').get_counts()
    local status = string.format('E:%d W:%d', counts.errors, counts.warnings)
<

================================================================================
TROUBLESHOOTING                                         *mtlog-troubleshooting*

Analyzer not found:
1. Install: `go install github.com/willibrandon/mtlog/cmd/mtlog-analyzer@latest`
2. Ensure it's in PATH: `which mtlog-analyzer`
3. Or set explicit path in config: `analyzer_path = '/full/path/to/mtlog-analyzer'`

No diagnostics appearing:
1. Check status: `:MtlogStatus`
2. Run manually: `:MtlogAnalyze`
3. Check kill switch: `:MtlogToggleDiagnostics`
4. Review suppressions: `:MtlogShowSuppressions`
5. Test context rules: `:MtlogContext test`

LSP integration issues:
1. Ensure gopls is installed: `go install golang.org/x/tools/gopls@latest`
2. Check LSP clients: `:LspInfo`
3. Verify config: `lsp_integration = { enabled = true }`
4. Try manual action: `:MtlogCodeAction`

Performance issues:
1. Adjust debounce: `debounce_ms = 500`
2. Limit concurrency: `queue = { max_concurrent = 2 }`
3. Disable auto-analyze: `auto_analyze = false`
4. Use context rules to limit scope

================================================================================
ABOUT                                                            *mtlog-about*

mtlog.nvim is part of the mtlog ecosystem, bringing structured logging
best practices to Go development.

Repository: https://github.com/willibrandon/mtlog
Issues: https://github.com/willibrandon/mtlog/issues
License: MIT

Related projects:
• mtlog - The core Go logging library
• mtlog-analyzer - Static analysis tool
• VS Code extension - For Visual Studio Code users
• GoLand plugin - For JetBrains IDE users

================================================================================
vim:tw=80:ts=4:ft=help:norl: