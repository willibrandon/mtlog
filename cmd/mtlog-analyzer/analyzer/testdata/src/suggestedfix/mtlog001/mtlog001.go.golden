package mtlog001

import "github.com/willibrandon/mtlog"

func testTemplateMismatchFixes() {
	log := mtlog.New()

	// Too few arguments - should suggest adding placeholders
	log.Information("User {UserId} performed {Action} on {Resource}", nil, nil, nil) // want `\[MTLOG001\] template has 3 properties but 0 arguments provided`
	// TODO: provide value for UserId, Action, Resource
	
	// Too few arguments with existing comment - TODO should go on next line
	log.Information("Failed {Operation}", nil) // existing comment // want `\[MTLOG001\] template has 1 properties but 0 arguments provided`
	// TODO: provide value for Operation

	// Too many arguments - should suggest removing extras
	log.Information("Hello {Name}", "Alice") // want `\[MTLOG001\] template has 1 properties but 3 arguments provided`

	// Correct number of arguments - no fix needed
	log.Information("User {UserId} logged in", 123)
}